### PR TITLE
MAINT: lint: enable line-length check (package-wide)

### LINKED ISSUE #19479: STY: comply with the documented line length limit
### Is your feature request related to a problem? Please describe.

Following gh-19461, it would help reviewers if CI could catch when lines exceed 88 characters.

We can't enable this in CI right now, since linting is done on whole files at once, so CI would fail on PRs due to violations in untouched lines. This is undesirable since the docs say that:

> It is recommended to leave existing style issues alone unless they exist in lines of code you are already modifying. This practice ensures that the codebase is gradually cleaned up without dedicating precious review time to style-only cleanups.

### Describe the solution you'd like.

To reach a point where we can enable the check in CI, we will need to deal with the ~2500 current violations (the number is so large mainly due to `scipy/constants/_codata.py` - adding a few `noqa`s there reduces the number to about 1000). I propose that we do so by:

1. Reducing line length or adding line breaks where appropriate.
2. Otherwise, adding `# noqa: E501` where we allow lines to exceed the limit.

I think that this should be done one submodule at a time. Perhaps a candidate for `'good first issue'` (cleaning up one submodule)?

### Describe alternatives you've considered.

1. Do nothing. Feel free to close this issue if anyone feels strongly for this option. This would require no extra immediate effort and may be preferable if the time to review PRs for this just doesn't seem worth it. This option will, however, inevitably lead to more violations of the limit creeping in as it is unrealistic to expect reviewers to check every line of every PR for violations manually.
2. Have one big PR to deal with them all at once. It might be easier to keep it isolated to one PR although the diff could get scarily large.

### Additional Information

- Not every violation would have to be dealt with - some of them occur in vendored git submodules, and others could be left alone if it is unlikely that any PRs will be touching that file any time soon. I would estimate that there are about 500 separate places which need a change, since many of the violations are the same structure across multiple lines or in files that won't be modified regularly.
- `# noqa: E501` is already used 18 times in SciPy.

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
Closes gh-19479

#### What does this implement/fix?
- Remove all undesired ignores for the line-length check package-wide
- So that the above can happen, clean up all existing violations
- Clean up `per-file-ignores`
- Also `noqa`s one `UP` error which ruff is picking up but mypy doesn't like my fix for

#### Additional information

- I'll write to the mailing list once noise has died down from release prep
- We can squash commits and add a git blame-ignore when reviewers are happy

### CODE REVIEW

File: scipy/sparse/_generate_sparsetools.py [hunk 1]
```diff
@@ -419,12 +419,15 @@ def main():
...
+                        f"_method(PyObject *, PyObject *);\n")
 
     method_struct = """\nstatic struct PyMethodDef sparsetools_methods[] = {"""
     for name in names:
-        method_struct += """
-        {{"{name}", (PyCFunction){name}_method, METH_VARARGS, NULL}},""".format(**dict(name=name))
+        method_struct += ("""
+            {{"{name}", (PyCFunction){name}_method, METH_VARARGS, NULL}},"""
+            .format(**dict(name=name))
+        )
```
Comment (lucascolley) [inline scipy/sparse/_generate_sparsetools.py:430]: I'm suspicious of these changes as the `Compiling C++ object scipy/sparse/sparsetools/_sparsetools.cpython-311-x86_64-linux-gnu.so.p/csr.cxx.o` step of `python dev.py build` seems to be taking abnormally long.

File: tools/lint.toml [hunk 2]
```diff
@@ -23,18 +23,11 @@ target-version = "py39"
 "scipy/stats/tests/data/_mvt.py" = ["E701", "E702"]
 # E501 (line length) ignores
 "benchmarks/benchmarks/**" = ["E501"]
-"scipy/_lib/**" = ["E501"]
```
Comment (lucascolley) [inline tools/lint.toml:26]: I assume that the CI linter does not complain about violations in git submodules as I have never seen a fail for them.

**Comment by lucascolley** [comment]
Just `optimize` and `special` left as the biggest offenders (+ `benchmarks` and `tools`). I'll write to the mailing list once noise has died down from release prep. Once this is ready we can squash commits and add a git blame-ignore.

File: scipy/stats/_continuous_distns.py [hunk 3]
```diff
@@ -1630,7 +1631,9 @@ def _isf(self, p):
         return -scu._cosine_invcdf(p)
 
     def _stats(self):
-        return 0.0, np.pi*np.pi/3.0-2.0, 0.0, -6.0*(np.pi**4-90)/(5.0*(np.pi*np.pi-6)**2)
+        m2 = (np.pi * np.pi / 3.0) - 2.0
+        m4 = -6.0 * (np.pi**4 - 90) / (5.0 * (np.pi * np.pi - 6)**2)
+        return 0.0, m2, 0.0, m4
```
Comment (lucascolley) [inline scipy/stats/_continuous_distns.py:1636]:
```suggestion
        v = (np.pi * np.pi / 3.0) - 2.0
        k = -6.0 * (np.pi**4 - 90) / (5.0 * (np.pi * np.pi - 6)**2)
        return 0.0, v, 0.0, k
```

File: scipy/optimize/_linprog_ip.py [hunk 4]
```diff
@@ -990,7 +990,8 @@ def _linprog_ip(c, c0, A, b, callback, postsolve_args, maxiter=1000, tol=1e-8,
 
     1. ``sksparse.cholmod.cholesky`` (if scikit-sparse and SuiteSparse are installed)
 
-    2. ``scipy.sparse.linalg.factorized`` (if scikit-umfpack and SuiteSparse are installed)
+    2. ``scipy.sparse.linalg.factorized``
+        (if scikit-umfpack and SuiteSparse are installed)
```
Comment (lucascolley) [inline scipy/optimize/_linprog_ip.py:994]: need to check that this renders properly

File: scipy/special/__init__.py [hunk 5]
```diff
@@ -768,7 +768,7 @@
    exprel    -- Relative error exponential, (exp(x)-1)/x, for use when `x` is near zero.
    sinc      -- Return the sinc function.
 
-"""
+"""  # noqa: E501
```
Comment (stefanv) [inline scipy/special/__init__.py:771]: Probably want to add these exceptions to the pyproject.toml configuration (all __init__ files are weird).

Comment (stefanv) [reply scipy/special/__init__.py:771]: example: https://github.com/scikit-image/scikit-image/blob/main/pyproject.toml#L180

**Review by stefanv — APPROVED** [review]
Thanks @lucascolley, looks good to me.

The only comment I had was that you may want to configure some general ignores for `__init__.py` files, instead of using `noqa`. Not a big deal.

I haven't checked every single line of the PR, but enough to convince myself that you've been careful :) Thank you!

**Comment by lucascolley** [comment]
I guess we want to squash merge at this point, then I can do the blame-ignores for this and gh-19623 in a follow-up?

**Comment by stefanv** [comment]
Thanks @lucascolley. Please go ahead.

**Pull request merged**